Added version-Information prepare-Script

luggesexe opened the pull request:
This commit adds an npm prepare command which fetches the current build-version from the angular-frontend and electron App.
This is needed for another PR where this information is displayed directly in the Electron App. This PR only displays the information while building and in the version.json file.

AKMaily commented:
Hello, 
thank you very much for your contribution ! I like the general idea of having the build versions for support as this could help a lot with user questions. 

Could you please provide a brief description of how to test the implemented behavior, along with the expected results ? This would be very helpful to review the effects of the code changes.

AKMaily reviewed (changes requested):
Hey, thank you for the update . I tested the feature with the userflow you described and everything worked fine. The code also looks good from my perspective.  

I do have a few suggestions regarding the documentation and consistency across the build scripts to make sure everyone can use the feature with the given documentation. Also, we're currently introducing a CHANGELOG.md to document changes in the application (see #41 ). It would be great if you could add your changes in a CHANGELOG.md as well.

luggesexe commented on package.json:
```diff
@@ -3,11 +3,12 @@
     "install:angular": "cd angular-frontend && npm ci",
     "install:electron": "cd electron && npm ci",
-    "build": "npm run build:angular && npm run build:electron",
+    "build": "npm run prepare && npm run build:angular && npm run build:electron",
```
I added and simplified the README.md with the new instructions provided by the repository.

AKMaily commented on .gitignore:
```diff
@@ -13,6 +13,7 @@ angular-frontend/npm-debug.log
 angular-frontend/yarn-error.log
 
 # IDEs and editors
+.idea
```
Apologies for the misunderstanding. I didn’t mean to suggest that you should create a new PR for this line of code.
I’m just trying to understand where it comes from, as it’s included in this PR but I currently don’t see how it is connected to the other changes made here.

AKMaily reviewed (changes requested):
With these changes, the README now explains how to build the project from the root directory.
However, running the build commands inside the individual sub-folders (where no explicit version is set) produces a different artifact than building from the root (where the version is defined).
Because the package.json includes several build commands, imho simply documenting the “preferred” command isn’t sufficient—the other commands still compile successfully and generate variants of the application with different feature sets.

I would suggest to either delete the other build options or include the feature for the other build options as well. The second suggestion would be my favorite. @tabeatheunicorn what is your opinion on this ?

luggesexe commented:
Hello @AKMaily,
I updated the requested changes to the README file and also changed the start:dev to start:angular. This makes it more clear which project is now built. I personally would prefer, if we would build the project from the root. imho there is no case, where we would want to build the angular frontend only or the electron only. This leaves one part of the project always outdated therefore I would do it by root dir.

AKMaily commented:
> Hello @AKMaily, I updated the requested changes to the README file and also changed the start:dev to start:angular. This makes it more clear which project is now built. I personally would prefer, if we would build the project from the root. imho there is no case, where we would want to build the angular frontend only or the electron only. This leaves one part of the project always outdated therefore I would do it by root dir.

Hi @luggesexe , 

thank you for the updates. I agree with you that building the project from the root directory is the easiest and most used way to build the project. However if the package.json files provide different ways for a developer to build the project, someone will. I fear that when we have inconsistent build options available in the code there will be problems with the versioning of the software and testing, as it is not clear which build process the developer used. 

Imo there are two options for this: 
1. Different build options exist, all lead to consistent build artifacts
2. Only one build process exisits and everybody has to follow one described build process 

As the root file now provides all build options and start options for the angular and electron application, deleting the other build scripts, would be the easiest way imo. 
@tabeatheunicorn i would like to have your opinion on this topic as you have more experience with software versioning and build processes later in the course of a larger software .

luggesexe commented:
I would use 2. (Only one build process exisits and everybody has to follow one described build process). We would have to double implement the same build procedure (as in in the electron folder we would have to go back with cd ../ and then run again). We would have to remember, to change all the package.json's if one location is changed. This could be forgotten and seems inefficient to me.

AKMaily commented on angular-frontend/src/app/graph/graph-data.service.ts:
```diff
@@ -4,18 +4,19 @@ import { line as d3Line } from 'd3-shape';
   private readonly $graphDimensions = signal({ width: 800, height: 600 });
-  private readonly $xDomain = signal([new Date(2020), new Date()]);
-  private readonly $yDomain = signal([0, 100]);
+  private readonly $xDomain = signal([new Date(0), new Date(0)]); 
```
These changes should not be included in this PR as they have nothing to do with the description i think they are part of #37

AKMaily reviewed (changes requested):
Merging the master-dev branch into this one seems to add features into this branch that are not included in the description. 

My suggestion would be the following: 

- revert the merge with the master-dev branch and only keep the commits necessary for this PR 
- delete the build scripts in the subfolders: angular_frontend and electron, keep the build scripts in the package.json in the root directory to maintain consistency 
- add your changes to the CHANGELOG.md file from the masterbranch

tabeatheunicorn commented on .gitignore:
```diff
@@ -137,4 +138,5 @@ electron/.vite/
-electron/res/angular
\ No newline at end of file
+electron/res/angular
+electron/src/version.js
```
```suggestion
# Is auto-generated by electron/src/version-information.js, so part of build artefact
electron/*/version.json
```

tabeatheunicorn commented on CHANGELOG.md:
```diff
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add Icon (#19)
 - Adding compodoc (#23)
 - Adding changelog information to the Contribution.md (#41)
+- Added version-Script to obtain package.json-Information (#35)
```
```suggestion
- Add version-script to obtain package.json information (#35)
```

tabeatheunicorn commented on CHANGELOG.md:
```diff
@@ -40,4 +41,5 @@ Formatting needs to be similar to the dataformat of the old OmnAIView data expor
-- Deletetion of deprecated Angular 18 patterns (#18)
\ No newline at end of file
+- Deletetion of deprecated Angular 18 patterns (#18)
+- Removed package.json build scripts in subfolders (#35)
```
```suggestion
- Remove `package.json` build scripts in subfolders (#35)
```

tabeatheunicorn commented on angular-frontend/package.json:
```diff
@@ -4,7 +4,6 @@
   "scripts": {
     "ng": "ng",
     "start": "ng serve",
-    "build": "ng build",
```
I don't see the need to remove the build and make scripts as they can be referenced from the top project as well.

AKMaily replied:
Referencing them from the top project with ```build:angular``` and running ```ng build``` in the subfolder lead to two applications, one with version information and one without. 
I would expect a command to always have the same output when the parameters are the same e.g. using build:angular or ng build in the angular_folder both have the framework_parameter "angular" and should build the same angular frontend. 

Imo build commands should be consistent and therefore deleting the inner build scripts is the correct way.

luggesexe replied:
@AKMaily Can you check, if this is as you expect it?

tabeatheunicorn commented on package.json:
```diff
@@ -3,14 +3,15 @@
-    "build:electron": "cd electron && npm run make",
-    "start:dev": "cd angular-frontend && npm run start",
+    "build": "npm run prepare && npm run build:angular && npm run build:electron",
+    "build:angular": "cd angular-frontend && ng build",
+    "build:electron": "cd electron && tsc && electron-forge make",
+    "start:angular": "cd angular-frontend && npm run start",
```
I would prefer to use npm features for this and contain the build and start instructions in the respective subfolder.
```
{
  "workspaces": [
    "angular-frontend",
    "electron"
  ],
  "scripts": {
    "build:angular": "npm run build --workspace=angular-frontend",
    "start:angular": "npm run start --workspace=angular-frontend"
  }
}
```

https://docs.npmjs.com/cli/v11/using-npm/workspaces

AKMaily replied:
@tabeatheunicorn your answers on this would be greatly appreciated as the changes of this PR currently block a working master version .

tabeatheunicorn commented on version-information.js:
Please add the intention of this file as documentation so that future developers understand when and why to use it

tabeatheunicorn commented on version-information.js:
```diff
@@ -0,0 +1,19 @@
+const fs = require('fs');
+const path = require('path');
+
+const electronPkgPath = path.join(__dirname, 'electron', 'package.json');
+const angularPkgPath = path.join(__dirname, 'angular-frontend', 'package.json');
+
+const electronPkg = JSON.parse(fs.readFileSync(electronPkgPath, 'utf-8'));
+const angularPkg = JSON.parse(fs.readFileSync(angularPkgPath, 'utf-8'));
```
We should add error handling here in case the directory structure changes. Also, maybe we find options to reduce the dependency from the project structure.

This is a future todo

tabeatheunicorn reviewed (approved):
I would have left the scripts in the subfolders as that makes the workspaces a little bit more intuitive. However, none of that is unchangeable so I am fine with the current format and updates of the Changelog as well.

Note that I did not verify whether this actually builds.

AKMaily reviewed (changes requested):
Thank you for the update. I tested the build and got a few more comments as it does not build in the current state and some parts of the documentation are not complete. 

Additionally I found an error regarding #36 . Deleting the brackets {} around MakerSquirrel in the forge.config.ts did mess up the forge runner load via jiti. Those brackets need to be added again otherwise it does not build correctly 

My suggestions would be the following : 

[ ] Update the documentation to fit the build process 
[ ] Update the other requested changes from the comments 
[ ] Add the brackets

luggesexe commented on CHANGELOG.md:
```diff
@@ -41,4 +42,6 @@ Formatting needs to be similar to the dataformat of the old OmnAIView data expor
-- Deletion of duplicated AsyncAPI description for OmnAIBackend 
\ No newline at end of file
+- Deletion of duplicated AsyncAPI description for OmnAIBackend 
+- Remove `package.json` build scripts in subfolders (#35)
```
Resolved with ad65883.

luggesexe commented on package.json:
```diff
@@ -1,16 +1,29 @@
+    "compodoc:serve":           "npm exec -w angular-frontend compodoc -s",
+    "serve:ssr":                "node electron/res/angular/server/server.mjs",
+    "start:electron":           "npm exec -w electron tsc && npm exec -w electron electron-forge start",
+    "build:electron":           "npm exec -w electron tsc && npm exec -w electron electron-forge make",
```
Resolved with 96b122a.

luggesexe commented on package.json:
```diff
@@ -1,16 +1,29 @@
+    "publish:electron":         "npm exec -w electron electron-forge publish",
+    "build":                    "npm run prepare && npm run build:angular && npm run build:electron"
+  },
+  "dependencies": {
```
Done f6d122d

luggesexe commented on README.md:
````diff
@@ -1,37 +1,36 @@
 
 1. Clone the project 
 2. Run ``` npm ci ``` in the root directory of the project 
 3. Navigate into the angular folder 
````
ab4ac9b

luggesexe commented on README.md:
````diff
@@ -41,103 +40,46 @@ The project currently contains two separat parts: An angular app and an electron
+#### Building
 
-To build the project run :
+To build the full project you can do so by running ```npm run build```. 
````
ab4ac9b

luggesexe commented on README.md:
````diff
@@ -41,103 +40,46 @@ The project currently contains two separat parts: An angular app and an electron
-
-    The project was generated using angular v19.1.4.
-    For more information on using the Angular CLI, including detailed command references, visit the [Angular CLI Overview and Command Reference](https://angular.dev/tools/cli) page.
+If you want to run the electron application separately in development mode, you can do so by running ```npm run start:electron```.
````
ab4ac9b

luggesexe commented:
@AKMaily I applied these changes. Please check. The current CI build displays as failed, because the "npm run install:all" is not present anymore. For that, the CI should be changed.
